src/onefetch/repo.rs

```diff
@@ -1,17 +1,122 @@
-use crate::onefetch::{commit_info::CommitInfo, error::*};
-use git2::{BranchType, Repository, RepositoryOpenFlags, Status, StatusOptions, StatusShow};
+use crate::onefetch::{commit_info::CommitInfo, error::*, utils};
+use git2::{
+    BranchType, Commit, Repository, RepositoryOpenFlags, Status, StatusOptions, StatusShow,
+};
 use regex::Regex;
 use std::path::Path;
 
-pub struct Repo {
-    repo: Repository,
+pub struct Repo<'a> {
+    repo: &'a Repository,
+    logs: Vec<Commit<'a>>,
 }
 
-impl Repo {
-    pub fn new(repo_path: &str) -> Result<Self> {
-        let repo = Repository::discover(repo_path)?;
-        Ok(Self { repo })
+impl<'a> Repo<'a> {
+    pub fn new(repo: &'a Repository, no_merges: bool) -> Result<Self> {
+        let logs = Repo::get_logs(repo, no_merges)?;
+        Ok(Self { repo, logs })
     }
+
+    fn get_logs(repo: &'a Repository, no_merges: bool) -> Result<Vec<Commit<'a>>> {
+        let mut revwalk = repo.revwalk()?;
+        revwalk.push_head()?;
+        let logs: Vec<Commit<'a>> = revwalk
+            .filter_map(|r| match r {
+                Err(_) => None,
+                Ok(r) => {
+                    if let Ok(commit) = repo.find_commit(r) {
+                        if no_merges {
+                            let parents = commit.parents().len();
+                            if parents > 1 {
+                                return None;
+                            }
+                        }
+                        Some(commit)
+                    } else {
+                        None
+                    }
```

Comment:
Since this is basically converting a `Result` to an `Option` that has been filtered, I think this can be simplified with `Result::ok` and `Option::filter`.
Something like
```rust
repo.find_commit(r).ok().filter(|commit| !(no_merges && commit.parents().len() > 1))
```
Once again, didn't test the above code, though :slightly_smiling_face:

Reply:
Alright worked directly.